File: tools/file-formats/distribution/reader.ts
import * as fs from 'fs'
import * as path from 'path'
import { BinaryReader } from '../../utils/binary-reader'

export async function readDistributionFile(file: string){
  const tokens = path.basename(path.dirname(file)).split('_')
  const region = [Number(tokens[1]), Number(tokens[2])]
  const data = await fs.promises.readFile(file)
  const reader = new BinaryReader(data.buffer as any)

  let count = reader.readUInt16()
  const slices: string[] = readStringArray(reader, count)
  const variants: string[] = readStringArray(reader, count)

  count = reader.readUInt32()
  const indices: number[] = []
  const positions: number[][] = []
  for (let i = 0; i < count; i++) {
    indices.push(reader.readUInt16())
  }
  for (let i = 0; i < count; i++) {
    const y = reader.readUInt16()
    const x = reader.readUInt16()
    positions.push([x, y])
  }
  return {
    region,
    slices: slices,
    variants: variants,
    indices: indices,
    positions: positions,
  }
}

function readStringArray(r: BinaryReader, count: number) {
  const result: string[] = []
  for (let i = 0; i < count; i++) {
    const c = r.readInt8()
    const v = String.fromCharCode(...r.readInt8Array(c))
    result.push(v)
  }
  return result
}

File: tools/nw-kit/formats/distribution/reader.go
package distribution

import (
	"io"
	"nw-buddy/tools/nw-kit/nwfs"
	"nw-buddy/tools/nw-kit/utils"
	"path"
	"strconv"
	"strings"
)

type Record struct {
	Region     [2]uint32   `json:"region"`
	Slices     []string    `json:"slices"`
	Variants   []string    `json:"variants"`
	Indices    []uint16    `json:"indices"`
	Positions  [][2]uint16 `json:"positions"`
	Positions2 [][2]uint16 `json:"positions2"`
	Types2     []byte      `json:"types2"`
	Positions3 [][2]uint16 `json:"positions3"`
	Types3     []byte      `json:"types3"`
}

func Load(f nwfs.File) (*Record, error) {
	data, err := f.Read()
	if err != nil {
		return nil, err
	}
	return Parse(data, f.Path())
}

func Read(r io.Reader, file string) (*Record, error) {
	data, err := io.ReadAll(r)
	if err != nil {
		return nil, err
	}
	return Parse(data, file)
}

func Parse(data []byte, file string) (res *Record, err error) {
	defer utils.HandleRecover(&err)

	res = &Record{}
	tokens := strings.SplitN(path.Base(path.Dir(file)), "_", 3)
	if len(tokens) == 3 {
		r1, _ := strconv.Atoi(tokens[1])
		r2, _ := strconv.Atoi(tokens[2])
		res.Region = [2]uint32{uint32(r1), uint32(r2)}
	}

	r := utils.NewByteReaderLE(data)
	count := int(r.MustReadUint16())
	if count > 0 {
		res.Slices = readStringArray(r, count)
		res.Variants = readStringArray(r, count)
	}

	count = int(r.MustReadUint32())
	res.Indices = make([]uint16, count)
	res.Positions = make([][2]uint16, count)
	if count > 0 {
		for i := 0; i < int(count); i++ {
			res.Indices[i] = r.MustReadUint16()
		}
		for i := 0; i < int(count); i++ {
			res.Positions[i] = [2]uint16{
				r.MustReadUint16(),
				r.MustReadUint16(),
			}
		}
		r.SeekRelative(int(count * 4))
		r.SeekRelative(int(count))
	}

	count = int(r.MustReadUint32())
	res.Positions2 = make([][2]uint16, count)
	res.Types2 = make([]byte, count)
	if count > 0 {
		for i := 0; i < int(count); i++ {
			res.Positions2[i] = [2]uint16{
				r.MustReadUint16(),
				r.MustReadUint16(),
			}
		}
		for i := 0; i < int(count); i++ {
			res.Types2[i] = r.MustReadByte()
		}
	}

	count = int(r.MustReadUint32())
	res.Positions3 = make([][2]uint16, count)
	res.Types3 = make([]byte, count)
	if count > 0 {
		for i := 0; i < int(count); i++ {
			res.Positions3[i] = [2]uint16{
				r.MustReadUint16(),
				r.MustReadUint16(),
			}
		}
		for i := 0; i < int(count); i++ {
			res.Types3[i] = r.MustReadByte()
		}
	}

	return res, nil
}

func readStringArray(r *utils.ByteReader, count int) []string {
	res := make([]string, count)
	for i := 0; i < count; i++ {
		c := r.MustReadByte()
		res[i] = string(r.MustReadBytes(int(c)))
	}
	return res
}
